processor/groupbytraceprocessor/storage_memory.go

```diff
@@ -0,0 +1,101 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package groupbytraceprocessor
+
+import (
+	"errors"
+	"sync"
+
+	"go.opentelemetry.io/collector/consumer/pdata"
+)
+
+var errNilResourceSpans = errors.New("the provided trace is invalid (nil)")
+
+type memoryStorage struct {
+	sync.RWMutex
+	content map[string]pdata.ResourceSpans
+}
+
+var _ storage = (*memoryStorage)(nil)
+
+func newMemoryStorage() *memoryStorage {
+	return &memoryStorage{
+		content: make(map[string]pdata.ResourceSpans),
+	}
+}
+
+func (st *memoryStorage) createOrAppend(traceID pdata.TraceID, trace pdata.ResourceSpans) error {
+	if trace.IsNil() {
+		return errNilResourceSpans
+	}
+
+	sTraceID := traceID.String()
+
+	st.Lock()
+	existing, ok := st.content[sTraceID]
+	if ok {
+		// we have a trace already, append the spans
+		for i := 0; i < trace.InstrumentationLibrarySpans().Len(); i++ {
+			spans := trace.InstrumentationLibrarySpans().At(i)
+			existing.InstrumentationLibrarySpans().Append(&spans)
+		}
+	} else {
+		newTrace := pdata.NewResourceSpans()
+		newTrace.InitEmpty()
+		trace.CopyTo(newTrace)
+		st.content[sTraceID] = newTrace
+	}
+	st.Unlock()
+
+	return nil
+}
+func (st *memoryStorage) get(traceID pdata.TraceID) (pdata.ResourceSpans, error) {
+	sTraceID := traceID.String()
+
+	st.RLock()
+	trace, ok := st.content[sTraceID]
+	if !ok {
+		return pdata.NewResourceSpans(), nil
+	}
+
+	result := pdata.NewResourceSpans()
```

Comment:
Why do you need to make a copy? Preventing potential mutations?
Would be more efficient to just return `trace` and leave a comment that you don't expect any mutations to happen outside.

Reply:
It's more efficient, but a future implementation (in disk storage) will have a different behavior and I prefer to have consistent behavior across implementations.